libcpuid/recog_amd.c

```diff
@@ -442,85 +417,79 @@ static int amd_has_turion_modelname(const char *bs)
 
 static struct amd_code_and_bits_t decode_amd_codename_part1(const char *bs)
 {
-	amd_code_t code = NC;
-	uint64_t bits = 0;
+	amd_code_t code = (amd_code_t) NC;
 	struct amd_code_and_bits_t result;
+	uint64_t bits = 0;
+	int i = 0;
 
-	if (strstr(bs, "Dual Core") ||
-	    strstr(bs, "Dual-Core") ||
-	    strstr(bs, " X2 "))
-		bits |= _X2;
-	if (strstr(bs, " X4 ")) bits |= _X4;
-	if (strstr(bs, " X3 ")) bits |= _X3;
-	if (strstr(bs, "Opteron")) bits |= OPTERON_;
-	if (strstr(bs, "Phenom")) {
-		code = (strstr(bs, "II")) ? PHENOM2 : PHENOM;
+	const struct { amd_code_t c; const char *search; } code_matchtable[] = {
+		{ PHENOM2, "Phenom(tm) II" },
+		{ PHENOM, "Phenom(tm)" },
+		{ FUSION_C, "C-##" },
+		{ FUSION_E, "E-###" },
+		{ FUSION_Z, "Z-##" },
+		{ FUSION_EA, "[EA]#-####" },
+	};
+
+	const struct { uint64_t bit; const char *search; } bit_matchtable[] = {
+		{ _X2, "Dual[- ]Core" },
+		{ _X2, " X2 " },
+		{ _X3, " X3 " },
+		{ _X4, " X4 " },
+		{ OPTERON_, "Opteron" },
+		{ ATHLON_, "Athlon" },
```

Comment:
It was `Athlon(tm)`, but it's `Athlon` since Ryzen (like AMD Athlon 200GE).

Reply:
Agreed, matching fror "Athlon(tm)" was needlessly specific.